### PR TITLE
fix(cursor): capture manually attached skills

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## Summary

- detect Cursor-native skills from `<manually_attached_skills>` on the matched user transcript row
- synthesize a paired Read tool call/result when Cursor inlines the skill without emitting Read hooks
- preserve transcript Read detection and avoid duplicate Windows records
- match the explicit `<user_query>` when selecting the turn

## Tests

- `npm test -- --run tests/unit/hooks/cursor/skill-detector.test.mjs tests/unit/hooks/cursor/inject-skill-records.test.mjs`
- 2 test files, 28 tests passed

### CODE REVIEW

File: assets/hooks/cursor-hook-processor.mjs [hunk 1]
```diff
@@ -364,10 +366,19 @@ async function main() {
...
-          // The Windows transcript assembler already materializes transcript
-          // tool_use entries. Only compensate paths assembled from hook events.
-          if (detectedSkills && detectedSkills.length > 0 && !assembledFromTranscript) {
-            injectSkillRecords(records, detectedSkills, runtimeConfig);
+          if (detectedSkills && detectedSkills.length > 0) {
+            // The Windows transcript assembler already materializes transcript
+            // Read tool_use entries. Always synthesize Read for explicit manual
+            // attachments; compensate actual transcript reads only on hook-event
+            // assembly paths so Windows does not emit duplicate Read records.
+            const readSkills = detectedSkills.filter(skill =>
```
Comment (linrunqi08) [inline assets/hooks/cursor-hook-processor.mjs]: **[Medium]** Windows(`assembledFromTranscript=true`)下,同时命中 `manual_attachment` 与 `transcript_read` 的 skill 会因 `manual_attachment` 通过此过滤并被 `injectSkillRecords` 合成一条 Read;但 transcript 组装器(`alignSteps`)此前已把该真实 `Read tool_use` 物化为 tool.call/result,导致同一 SKILL.md 在同一轮出现**重复 Read 记录**。

**影响:** 违反本段注释声称的 "so Windows does not emit duplicate Read records";下游按 Read/skill 计数会重复计数,且重复记录带独立 event.id/tool.call.id,难以去重。

**建议:** 在 `assembledFromTranscript` 分支只合成"纯手动附加(不含 transcript_read)"的 skill:
```js
const readSkills = detectedSkills.filter(skill => {
  const s = skill.detectionSources || [];
  if (!assembledFromTranscript) {
    return s.includes("manual_attachment") || s.includes("transcript_read");
  }
  // Windows: 真实 Read 已被 assembler 物化,仅对纯手动附加合成
  return s.includes("manual_attachment") && !s.includes("transcript_read");
});
```

*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/cursor/skill-detector.mjs [hunk 2]
```diff
@@ -35,46 +47,60 @@ export function detectSkillFromTranscript(transcriptPath, userPrompt) {
...
 
   let matchedTurnStart = -1;
+  let matchedUserText = '';
   for (let i = 0; i < entries.length; i++) {
     const e = entries[i];
     if (e.role !== 'user') continue;
     const userText = extractUserText(e);
-    if (userText && normalizeForMatch(userText).includes(normalizedPrompt)) {
+    const userQuery = extractUserQuery(userText);
+    const normalizedCandidate = normalizeForMatch(userQuery || userText);
```
Comment (linrunqi08) [inline assets/hooks/cursor/skill-detector.mjs:62]: **[Medium]** prompt 匹配由旧的全文 `includes` 收紧为"存在 `<user_query>` 时要求完全相等"(`normalizedCandidate === normalizedPrompt`)。当 hook 侧 `beforeSubmitPrompt.prompt` 与 transcript 的 `<user_query>` 归一化后仍不完全一致(例如 Cursor 追加选区/文件引用或改写 prompt)时,会直接 `return null`,导致该轮 skill 检测整体失效(静默漏采),而旧实现凭 `includes` 仍可命中。

**影响:** transcript_read 主路径在 prompt 与 user_query 存在细微差异的真实场景下可能漏采,属静默数据丢失且不易发现。

**建议:** 保留"用 user_query 作匹配基准"的安全性,但把相等放宽为对 user_query 的包含匹配:
```js
const base = userQuery || userText;
const normalizedCandidate = normalizeForMatch(base);
if (normalizedCandidate === normalizedPrompt ||
    normalizedCandidate.includes(normalizedPrompt)) { ... }
```
("prefer user_query" 测试中 user_query 不含另一 prompt,仍返回 null,用例不受影响。)

*Generated by LoongSuite-Pilot Code Review Agent*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 2 |
| Low      | 2 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

本 PR 为无状态 best-effort hook：仅新增同步 `fs.readFileSync` 与纯函数解析，无定时器/watcher/子进程；检测逻辑整体包在 `try/catch` 中，异常不阻塞输出。

### 总体结论

无阻断合入项。方向正确、测试充分（28/28 通过）。**建议合入前**处理两个 Medium：

1. **Windows 重复 Read（Medium）**：`assembledFromTranscript=true` 时，同时命中 `manual_attachment` + `transcript_read` 的 skill 会被合成一条 Read，而 transcript 组装器已物化了真实 Read → 同一 SKILL.md 在同一轮出现重复 Read 记录，违反注释声称的 "Windows does not emit duplicate Read records"，下游会重复计数。修复：在 `assembledFromTranscript` 分支只对"纯手动附加（不含 transcript_read）"合成。
2. **prompt 精确匹配收紧致漏采（Medium）**：存在 `<user_query>` 时由旧的全文 `includes` 收紧为完全相等；当 hook prompt 与 transcript user_query 归一化后仍有细微差异时会整轮 `return null` 静默漏采。建议放宽为对 user_query 的 `includes` 匹配（既保留防内联正文污染的安全性，又不破坏现有用例）。

另有 2 个 Low：`extractManuallyAttachedSkills` 会扫描内联 SKILL.md 正文（与注释不符，存在误报向量）；缺少处理器级 both-source 去重回归用例。

### Highlights（正向实践）

- `Map(skillName+skillPath)` 去重并合并 `detectionSources`，`manual_attachment` 优先作为主 source，语义清晰。
- 新增 `block.input?.file_path` 兼容 `ReadFile` 变体入参。
- 用 `<user_query>` 作为匹配基准防内联正文污染，方向正确。
- 保留 `transcript_post_assembly` 兜底，向后兼容良好；单测覆盖面广。

---
*评审报告详见: `code-review/pr-193/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by ralf0131 — APPROVED** [review]
LGTM. Clean implementation of manual skill attachment detection.

The distinction between `manual_attachment` and `transcript_read` detection sources is well-designed — it correctly handles the case where Windows already materializes transcript tool_use entries while hook-event assembly paths still need compensation. The `<manually_attached_skills>` parsing logic is robust with proper fallback for unquoted paths.

Test coverage is thorough, covering edge cases like missing files, empty sections, and mixed skill types.

---
*Automated review by github-manager-bot*

File: assets/hooks/cursor-hook-processor.mjs (hunk 1 above)
Comment (rangemer333-cell) [reply assets/hooks/cursor-hook-processor.mjs]: Windows transcript 组装路径只合成“纯 manual_attachment”的 Read。
同时包含 manual_attachment 和 transcript_read 时，不再重复合成。
补齐三组过滤测试。

File: assets/hooks/cursor/skill-detector.mjs (hunk 2 above)
Comment (rangemer333-cell) [reply assets/hooks/cursor/skill-detector.mjs:62]: 经验证，prompt 与 user_query不会存在差异，cursor会将同样的一段prompt落盘，不会出现改写prompt

File: assets/hooks/cursor-hook-processor.mjs (hunk 1 above)
Comment (linrunqi08) [reply assets/hooks/cursor-hook-processor.mjs]: ✅ **已确认修复。**

抽出的 `filterSkillsForReadInjection` 在 `assembledFromTranscript=true` 时改为 `manual_attachment && !transcript_read`,both-source skill 不再与 assembler 已物化的真实 Read 重复合成,逻辑正确。已拉取 `1354788` 逐行核对 + 运行单测(28→31 全过,含新增的 dual-source 过滤用例)。此项闭环。

*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/cursor/skill-detector.mjs (hunk 2 above)
Comment (linrunqi08) [reply assets/hooks/cursor/skill-detector.mjs:62]: 收到,接受该澄清 👍 —— prompt 与 `<user_query>` 由 Cursor 原样落盘、不改写,则精确相等匹配成立,此项按 false-positive 关闭。

一点残留提示(不阻塞合入):该正确性依赖 Cursor 落盘行为的一致性,并未由代码强制。若未来 Cursor 改变行为(例如向 prompt 追加选区/@mention 上下文,而 `<user_query>` 只存纯文本),精确相等会静默漏采。可留意,无需现在处理。

*Generated by LoongSuite-Pilot Code Review Agent*

**Review by linrunqi08 — APPROVED** [review]

**Review by ralf0131 — APPROVED** [review]
﻿### Summary
Well-crafted fix that adds detection of Cursor's `<manually_attached_skills>` metadata, distinguishing between `manual_attachment` and `transcript_read` evidence sources. The `filterSkillsForReadInjection` logic correctly avoids duplicating Read records on transcript-assembled paths.

### Highlights
- **Security-conscious prompt matching**: prefers explicit `<user_query>` extraction over fuzzy matching, preventing inlined skill content from accidentally satisfying prompt matches
- **Smart deduplication**: case-insensitive Map key with source tracking — dual-source skills correctly get both evidence types
- **Clean separation**: `filterSkillsForReadInjection` encapsulates the transcript-vs-hook-event assembly logic
- **Thorough tests**: manual attachment, dedup, multi-skill, repeated prompts all covered

LGTM.

---
*Automated review by github-manager-bot*
